File: extension/shared/prompt-transfer.js
(function (global) {
    'use strict';

    const CURRENT_SCHEMA_VERSION = 1;

    function parseImportText(text) {
        let parsed;

        try {
            parsed = JSON.parse(text);
        } catch (error) {
            return { prompts: [], issues: [], fatalError: 'The selected file is not valid JSON.' };
        }

        let records;
        let format;

        if (Array.isArray(parsed)) {
            records = parsed;
            format = 'legacy-array';
        } else if (parsed && typeof parsed === 'object' && Array.isArray(parsed.prompts)) {
            if (parsed.schemaVersion !== CURRENT_SCHEMA_VERSION) {
                return {
                    prompts: [],
                    issues: [],
                    fatalError: `Unsupported backup schema version: ${String(parsed.schemaVersion)}.`
                };
            }
            records = parsed.prompts;
            format = 'versioned-envelope';
        } else {
            return {
                prompts: [],
                issues: [],
                fatalError: 'Backup must be a prompt array or a supported versioned envelope.'
            };
        }

        const normalized = global.PromptSchema.normalizePromptCollection(records);
        return {
            prompts: normalized.prompts,
            issues: normalized.issues,
            fatalError: null,
            format
        };
    }

    function mergePromptCollections(existingPrompts, importedPrompts) {
        const existingIds = new Set(existingPrompts.map(prompt => prompt.id));
        const added = [];
        const conflicts = [];

        importedPrompts.forEach(prompt => {
            if (existingIds.has(prompt.id)) {
                conflicts.push(prompt.id);
                return;
            }
            existingIds.add(prompt.id);
            added.push(prompt);
        });

        return {
            prompts: existingPrompts.concat(added),
            added,
            conflicts
        };
    }

    function createExportEnvelope(prompts) {
        const normalized = global.PromptSchema.normalizePromptCollection(prompts);

        if (normalized.issues.length) {
            throw new Error('Cannot export a prompt library containing invalid or duplicate records.');
        }

        return {
            schemaVersion: CURRENT_SCHEMA_VERSION,
            exportedAt: new Date().toISOString(),
            prompts: normalized.prompts
        };
    }

    global.PromptTransfer = Object.freeze({
        CURRENT_SCHEMA_VERSION,
        parseImportText,
        mergePromptCollections,
        createExportEnvelope
    });
})(window);

File: launcher/PromptLauncher.Core/PromptTransfer.cs
using System.Text.Json;

namespace PromptLauncher.Core;

public static class PromptTransfer
{
    public const int CurrentSchemaVersion = 1;
    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    public static ImportResult Parse(string text, IReadOnlyList<PromptRecord>? existing = null)
    {
        JsonDocument document;
        try { document = JsonDocument.Parse(text); }
        catch (JsonException) { return new([], [], [], [], "The selected file is not valid JSON.", null); }

        using (document)
        {
            var root = document.RootElement;
            JsonElement records;
            string format;
            if (root.ValueKind == JsonValueKind.Array)
            {
                records = root;
                format = "legacy-array";
            }
            else if (root.ValueKind == JsonValueKind.Object && root.TryGetProperty("prompts", out records) && records.ValueKind == JsonValueKind.Array)
            {
                if (!root.TryGetProperty("schemaVersion", out var version) || version.ValueKind != JsonValueKind.Number || version.GetInt32() != CurrentSchemaVersion)
                {
                    var shown = root.TryGetProperty("schemaVersion", out version) ? version.ToString() : "undefined";
                    return new([], [], [], [], $"Unsupported backup schema version: {shown}.", null);
                }
                format = "versioned-envelope";
            }
            else return new([], [], [], [], "Backup must be a prompt array or a supported versioned envelope.", null);

            var normalized = PromptSchema.Normalize(records);
            var current = existing ?? [];
            var ids = current.Select(prompt => prompt.Id).ToHashSet(StringComparer.Ordinal);
            var added = normalized.Prompts.Where(prompt => ids.Add(prompt.Id)).ToList();
            var conflicts = normalized.Prompts.Where(prompt => !added.Contains(prompt)).Select(prompt => prompt.Id).ToList();
            return new(current.Concat(added).ToList(), added, conflicts, normalized.Issues, null, format);
        }
    }

    public static string Export(IReadOnlyList<PromptRecord> prompts, DateTimeOffset? exportedAt = null)
    {
        var normalized = PromptSchema.Normalize(prompts);
        if (normalized.Issues.Count > 0)
            throw new InvalidDataException("Cannot export a prompt library containing invalid or duplicate records.");

        return JsonSerializer.Serialize(new
        {
            schemaVersion = CurrentSchemaVersion,
            exportedAt = (exportedAt ?? DateTimeOffset.UtcNow).ToString("O"),
            prompts = normalized.Prompts.Select(ToJsonRecord)
        }, JsonOptions);
    }

    internal static object ToJsonRecord(PromptRecord prompt) => new
    {
        id = prompt.Id,
        title = prompt.Title,
        category = prompt.Category,
        description = prompt.Description,
        text = prompt.Text,
        pinned = prompt.Pinned
    };
}
